[#156] 글쓰기: Core Data 저장

jwonyLee opened the pull request:
## 개요

<img src="https://user-images.githubusercontent.com/15073405/140937105-ab595d9a-c02f-4e6b-8021-478b8b942ea3.gif" width="50%" />

사용자에게 입력받은 항목을 Core Data에 저장한다.

## 작업 사항

- 셀이 재활용될 때 setupBind() 호출
  - 이 부분은 제가 몰랐던 부분인데.. 다른 화면(앨범, 카테고리)으로 이동했다가 글쓰기 화면으로 돌아오면 셀이 다시 그려지면서 `disposeBag`을 초기화하는데요, 이 때 초기화되면서 기존에 구독했던 것들이 초기화되기 때문에 재호출하는 코드를 추가했습니다.
- 사용자에게 입력받은 항목을 Core Data에 저장
  - PostRepository 를 이용해 Core Data에 게시물을 저장하는 기능을 구현했습니다.
  - 필수 항목(사진)이 누락된 경우 사용자에게 알림
    - `selectedImages`가 비어있는 경우 `showRequiredAlert()`을 호출합니다.
    - `!selectedImages.isEmpty`가 코드를 읽을 때 부자연스러워서 `Collection.isNotEmpty`를 구현해 가독성을 늘리고자 했습니다.
- WriteViewController DataSource 리팩토링
  - cell의 값을 읽어오는 과정에서 `tableView(_:cellForRowAt:)` 메서드가 너무 비대해져서 셀 구성하는 코드를 분리했습니다.
- WriteType을 PostType으로 통일
  - WriteType과 PostType이 동일한 역할을 하고 있었기 때문에 PostType으로 통일했습니다.
  - 이 과정에서 기존에 WriteType에 있었던 `image: UIImage?` 프로퍼티를 PostType으로 옮겼습니다.
- ScoreType unrated 추가
  - 기획 상에서 필수 입력 항목이 사진뿐이기 때문에 만족도 미선택을 나타내는 unrated = 0 을 추가했습니다.

## 동작 방식
- WrtieViewController는 WrtieViewModel을 가지고 있다. 
- WriteViewModel
    - 글쓰기 화면에 필요한 데이터를 가지고 있다. Rx.Subject를 이용하여, 데이터의 변형을 다른 객체가 (WriteViewController, Wrtie**Cell 등) 감지 할 수 있도록 한다.
    - 데이터가 필요한 뷰는 WrtieViewModel의 Subject를 가지고 있다. 그렇기 때문에 WriteViewModel(혹은 뷰)에서 데이터가 변경되더라도 별도의 데이터 전달없이 동일하게 반영된다.
- CategoryViewController에서 Category들을 선택하고 [확인] 버튼을 통해 WriteViewController로 돌아오는 경우에 NotificationCenter를 이용하여 선택된 Category들을 post한다. 그리고 이를 WrtieViewModel이 구독한다. 
- PhotoViewController에서 선택한 이미지들을 NotificationCenter를 이용하여 post하고, WrtieViewModel이 구독하여 thumbnail, originImage를 만들어낸다.
- WriteImageTableCell
    - WriteViewModel의 thumbnailSubject를 주입받고, 구독하여 데이터를 표시한다. 
    - WriteImageTableCell 에서 이미지를 삭제하는 경우, NotificationCenter를 이용하여 삭제된 indexPath를 post한다. 이 Notification은 WriteViewModel이 구독하여 동일한 위치의 originalImage를 삭제한다.

## 예외 사항

- GUI 미적용 (별도로 진행할 예정)

### Linked Issue

close #136
close #156

jwonyLee committed:
[#156] fix: 셀이 재활용될 때 setupBind() 호출

다른 화면(앨범, 카테고리)으로 이동했다가 글쓰기 화면으로 돌아오면 셀이 다시 그려진다. 이 때 disposeBag이 초기화되면서 기존에 구독했던 것들이 풀리기 때문에 disposeBag을 초기화하고 나서 구독을 다시해줘야 한다.

jwonyLee committed:
[#156] ref: WriteViewController 디렉토리 분리
- RatingView.didTapButtonBlock?() 호출 시점 변경

jwonyLee committed:
[#136] ref: WriteType을 PostType으로 통일

jwonyLee committed:
[#156] feat: 사용자가 입력한 항목의 값을 추출, 반환하는 메서드 구현
- getTypeSectionValue() : Section.type 값 추출
- getRating() : Section.rating.currentRating 값 추출
- getContents() : Section.contents.text 값 추출

jwonyLee committed:
[#156] feat: Core Data에 새로운 게시물을 저장한다.

jwonyLee committed:
[#156] ref: WriteViewController TableView 리팩토링
- DataSource 별도 분리
- 셀 구성 메서드 분리

jwonyLee committed:
[#156] feat: 필수 항목이 누락되었을 때 사용자에게 알럿 표시
- Collection.isNotEmpty 구현
- selectedImage.isEmpty 일 때 showRequiredAlert() 호출
- showRequiredAlert() 구현

jwonyLee committed:
[#156] feat: ScoreType.unrated 추가
만족도 미기재를 위한 unrated case 추가

gustn3965 commented on ThingLog/Coordinator/WriteCoordinator.swift:
```diff
@@ -24,7 +24,7 @@ final class WriteCoordinator: Coordinator {
 
     /// WriteViewController로 이동한다.
     /// - Parameter type: WriteType에 따라 ViewModel을 생성하기 위해 필요하다.
```
타입이름이 변경됐으므로, 주석에서도 PageType으로 변경해야할 것 같습니다~

gustn3965 commented on ThingLog/ViewController/Write/WriteViewController.swift:
```diff
@@ -104,6 +105,7 @@ final class WriteViewController: BaseViewController {
         bindKeyboardWillShow()
         bindKeyboardWillHide()
         bindNotificationPassSelectPHAssets()
```
`WriteViewController`가 글쓰기 완료를 눌러 뷰컨트롤러가 보이지않아도 지금 메모리에 계속 남아있는 상태입니다.  ( 아래 사진은 5번 글쓰기 했을때, 모두 남아있는 상태입니다 ) 
![0x105f06540](https://user-images.githubusercontent.com/48749182/141088893-fc0cc34a-6b4c-46f6-bdc4-467381cd8afb.jpg)

의심갈만한 코드들을 주석처리해보면서 다음과같이, `bindNotificationPassSelectPHAsset()`에서 문제가 있는걸 확인할 수 있었습니다.

사실 기존코드도 큰 이상은 없다라고 생각되는데, `bind` 캡처리스트에서 사용하지 않고, `bind` 안의 `DispatchQueue`에서 캡처리스트 ( `weak self` ) 로만 해도 메모리가 해제되지 않았습니다. 그래서 `bind` 에서 직접 `weak self` 로 사용하니, 메모리에서 정상적으로 해제가 되는 걸 볼 수 있었습니다. 저도 왜 그런지 정확히 이유를 모르겠습니다. 😫 혹시 알게되시면 공유 해주시면 감사하겠습니다. 

```swift
    func bindNotificationPassSelectPHAssets() {
        NotificationCenter.default.rx
            .notification(.passSelectAssets, object: nil)
            .map { notification -> [PHAsset] in
                notification.userInfo?[Notification.Name.passSelectAssets] as? [PHAsset] ?? []
            }
            .bind {  [weak self]  assets in
                DispatchQueue.main.async {
                    guard let self = self else { return }
                    self.selectedImages = self.viewModel.requestThumbnailImages(with: assets)
                }
            }.disposed(by: disposeBag)
    }
```

jwonyLee replied:
중첩된 클로저가 있을 때 더 높은 수준의 클로저에서 `[weak self]`를 호출해야 하고, 내부 클로저에 추가하면 메모리 누수가 발생한대요. 덕분에 새로운 거 공부하고 갑니다. 감사합니당!
https://medium.com/flawless-app-stories/the-nested-closure-trap-356a0145b6d

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -73,6 +91,65 @@ final class WriteViewModel {
+                    attachments: attachments)
+    }
+
+    /// originalImages로 [Attachment] 를 생성한다.
+    private func createAttachment() -> [Attachment] {
+        let thumbnailSize: CGSize = CGSize(width: UIScreen.main.bounds.width / 3, height: UIScreen.main.bounds.width / 3)
+        var attachments: [Attachment] = []
+        for index in 0..<originalImages.count {
+            if let thumbnail: UIImage = originalImages[index].resizedImage(Size: thumbnailSize) {
+                let attachment: Attachment = Attachment(thumbnail: thumbnail,
+                                                        imageData: .init(originalImage: originalImages[index]))
+                attachments.append(attachment)
+            }
+        }
+        return attachments
+    }
 }
 
 extension WriteViewModel {
```
extension에 있는 `executeOriginalImages` 와 같이 코멘트를 남깁니다. ( 해당 코드는 코멘트를 달 수가 없더라구요. 변경사항이 없어서 그런지.. ) 

 디바이스에 저장할때,  `executeOriginalImages()` 메소드에서  `PHImageManagerMaximumSize` 로 저장하고 있는데, 
이때문에, 요즘 최신 핸드폰이나, 화질이 굉장히 좋은 경우에 너무많은 메모리를 차지하게 됩니다. 
특히나 제 폰 같은 경우에는, 카메라로 찍으면 한장에 `2MB` 정도 되는데,   `PHImageManagerMaximumSize` 로 변환하게 되면 20mb가 차지하게 됩니다. 
그리고 이때문에, 이러한 사진 10장을 저장하는 경우에 ( 작성완료 ) , 굉장히 오래걸리는 뿐만 아니라 굉장히 메모리를 많이 사용하고 있는걸  확인하셨을겁니다. 
![462 1 MB](https://user-images.githubusercontent.com/48749182/141110529-b03962d7-e703-483b-8d3c-b4c939eb3c71.jpg)
그러면, 10장에 총 200mb나 사용되므로, 이는 굉장히 메모리 낭비이면서, 사용자입장에서도 좋은 방법은 아닌 것 같습니다. 



타협안으로, 저는 `MaximumSize`보다는,  디바이스크기나, 그보다 작은 사이즈로 저장하는게 더 좋다라고 생각합니다. 

그리고, `thumbnail`사진도 대략 150 * 150사이즈로 나오게되는데, 이것도 좋은 2MB이미지에서는 약 0.3mb 보다 더 높게 나오므로, 이것도 썸네일 사진에 과도한 메모리라고 생각합니다. 그래서 썸네일에서는 딱 맞는 크기에 깨끗하게 나오는 것보다, 어느정도 깨져도 되니, 대신 메모리 측면을 아끼는 방법이 더 좋다라고 생각합니다. `50*50` 정도 해도 크게 불편한 이미지는 아니라고 생각합니다.

gustn3965 reviewed (commented):
고생하셨습니다!! 기능에 대해서는 크게 문제는 없는 것 같습니다!

일단, 전부 리뷰는 못했고, 테스트하는 도중에 메모리 이슈로 한 3시간은 쓴 것 같습니다. ㅎㅎ; 

그래서 우선 발견한 부분만 리뷰를 올리고 나중에 코드도 다시 보도록 하겠습니당 🙏🏻

jwonyLee committed:
[#156] ref: 피드백 반영
- WriteType -> PageType 변경에 따라 주석, 이니셜라이저 수정
- Attachment.toEntity(in:) 에서 pngData() 대신 jpegData(compressionQuality: 0.8)로 변경
- 중첩된 클로저에서 메모리 누수 부분 수정
- thumbnailSize, imageSize 변경

jwonyLee commented:
다음과 같이 수정했습니다. 
- WriteType -> PageType 변경에 따라 주석, 이니셜라이저 수정
- `Attachment.toEntity(in:)` 에서 pngData() 대신 jpegData(compressionQuality: 0.8)로 변경
  - 기존에는 png로 저장하고 있었는데, 이 부분을 jpeg로 0.8 압축해서 저장하는 것으로 변경했습니다. (썸네일, 이미지 둘 다)
- 중첩된 클로저에서 메모리 누수 부분 수정
- thumbnailSize, imageSize 변경
  - WriteViewController에서 보여주는 썸네일 크기와 메인 화면 등에서 보여주는 이미지 크기를 100 * 100으로 통일했습니다. 제안해주신 50 * 50은 iPhone 13 Pro Max 시뮬레이터에서 눈에 띄게 사진이 뭉개져보여서 그보다 조금 더 크게 설정할 수 밖에 없었습니다.
  - 포토카드로 내보낼 때의 크기를 고려해서 `MaximumSize` 대신 588 * 588 사이즈로 변경했습니다. 이 크기는 병연님께 포토 카드를 내보낼 때 적절한 이미지 크기를 여쭤봐서 정하게 되었습니다.

동일한 사진을 jpeg, 0.8로 압축을 해봤는데요, `PHImageManagerMaximumSize`일 때 1067435 bytes (1.01 MB), 588 * 588 일 때 127918 bytes (0.12 MB)로 대략 10배 정도가 차이가 나더라구요. 말씀하신 대로 MaximumSize는 여러가지 측면에서 좋지 못했던 선택인 거 같습니당 ;ㅅ;

gustn3965 commented on ThingLog/Model/PostType.swift:
```diff
@@ -7,13 +7,28 @@
 import CoreData
 import Foundation
+import UIKit.UIImage
 
```
저는 `PostType`과 `PageType`이 계속 햇갈려서, 또 그에 마땅한 네이밍도 떠오르지 않아서, 따로 제안할 의견은 없습니다. 
다만 주석에 `PostType`과 다른점과 필요성에 대해 명시하면 더 좋을 것 같습니다~! 👍

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -18,13 +18,28 @@ final class WriteViewModel {
         case type
         case rating
         case contents
+
+        func cellIdentifier() -> String {
+            switch self {
+            case .image:
+                return WriteImageTableCell.reuseIdentifier
+            case .category:
+                return WriteCategoryTableCell.reuseIdentifier
+            case .type:
+                return WriteTextFieldCell.reuseIdentifier
+            case .rating:
+                return WriteRatingCell.reuseIdentifier
+            case .contents:
+                return WriteTextViewCell.reuseIdentifier
+            }
+        }
     }
```
저는 이 코드가 지금은 편리하다라고 생각할 수 있지만, 장기적으로 봤을 때는 도움보다는 오히려 단점을 유발할 것 같습니다.
왜냐하면,  나중에 해당 섹션의 Cell타입이 변경된다면 기존의 Cell타입이 사용된 코드를 모두 새로운 Cell타입을 변경해주어야 하는데, 그럼 `WriteViewModel`의 `cellIndetifier()` 안에서도 변경해주어야 합니다.  즉, 하나의 변경에 대해서 다수의 변경이 발생하고 있습니다. 

또한, `WriteViewController` 의 `dequeueAndConfigureCell` 에서 이 메서드를 이용하여 `identifier`를 사용하면서, 또 다시 밑에서 `타입케스팅`을 진행하고 있습니다. 

그리고 이와는 별개로 `WriteViewController`의 `tableView`에 개별적으로, 필요한 `cell`들을 `register`메서드로 각각 등록하고 있습니다. 

종합적으로 봤을 때, `cellIndeitifer()` 메서드를 이용하여 얻는 장점이 `dequeueAndConfigureCell`에서 한번만 발생하지만, 이로인해 의존성은 더 추가됐습니다.

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -35,17 +50,22 @@ final class WriteViewModel {
+    private(set) var originalImages: [UIImage] = []
+    private var selectedCategories: [Category] = []
+    /// 이미지가 저장될 크기
+    private let imageSize: CGSize = CGSize(width: 588, height: 588)
```
이 프로퍼티는  `// MARK: Properties`로 이동해야할 것 같습니다.

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -120,10 +198,11 @@ extension WriteViewModel {
         options.isSynchronous = true
 
         DispatchQueue.global().async { [weak self] in
+            guard let self = self else { return }
```
`requestThumbnailImages` 이 메서드와 `executeOriginalImages` 의 동작 내용은 동일하다라고 생각됩니다. 하지만 메서드명은 하나는 `request`, 하나는 `execute`로  둘을 동일하게 할 필요가 있는 것 같습니다.

그리고  `requestThumbnailImages`에서 썸네일크기 이미지를 받아와 return하고, `createAttachment`메서드에서도 썸네일크기로 다시한번 리사이징 하는데,  총 두번의 리사이징을 하고 있으므로, 한번만 하는게 좋을 듯 싶습니다.

gustn3965 reviewed (commented):
고생하셨습니다 ~  반영하신 부분들 잘 확인했습니다 ~ 아래는 따로 달 수가 없어서 리뷰에 추가했습니다! 

1. 지금 `PhotosViewController`만 계속 들어갔다 나왔다를 반복하면 아래와 같이 메모리가 계속 증가하는 걸 볼 수 있습니다. 
![ThingLog PID 10953](https://user-images.githubusercontent.com/48749182/141670563-70b39b97-29d3-4a16-a7ad-d9f2635ecfd6.jpg)

그 이유를 나름대로 찾아보니, ( 확실한 건 아닙니다 ) `PHPhotoLibrary`가 캐싱을 하고 있으므로, 계속 메모리가 남아있는 현상이 발생하고 있다라고 판단했습니다. 
`PhotosViewController`를 deinit할 때, cache를 모두 비워주는 방법을 하니, 이전과 같이 크게 증가하지는 않았습니다.

```swift
    deinit {
        imageManager.stopCachingImagesForAllAssets()
        PHPhotoLibrary.shared().unregisterChangeObserver(self)
    }
```



2. 두번째로는, 사진을 추가한 상태에서 다시 삭제해도 삭제가 반영이 되지 않는 것 같습니다. 
`WriteImageTableCell` 에서 사진을 삭제할 때,  사진만 삭제하고, 별도로 `Notification`을 전달하거나, `WriteViewModel`에게 전달하는 로직이 없어서 이런 현상이 발생하는 듯 합니다.


3. 추가적으로,  `Notification.passSelectAssets` 을 살펴보니, 이것을 구독하는 객체가 두 개가 있는걸 확인했습니다 (`WriteViewController`, `WriteViewModel` )
`WriteViewController`는 이것을 구독하여,  `WriteImageTableCell` 를 다시 그리도록 하고 있고, 
`WriteViewModel`은 이것을 구독하여, `original Image`를 만들어내고 있습니다. 
지금 하나의 이벤트에 다수 옵저빙을 하고 있는데, 그 목적에 좋은 구조는 아닌 듯 합니다. 왜냐하면, `WriteViewModel`은` ViewModel`로써, 데이터로직을 담당하여 데이터를 전달하는 목적을 가져야할 텐데, 이를 지금 `WriteViewController`도 같이하고 있어, 코드의 복잡성과 의존성만 높이고 있다라고 생각됩니다. 
저는 모든 필요한 데이터, 관련 로직들이 `WriteViewModel`이 전담하도록 하는 편이 나을 것 같습니다.  `WriteViewModel`만 옵저빙하여, `WriteViewController`가 `WriteViewModel`에 대해 옵저빙하여 `tableView`를 `reload`하는 식으로 하는게 나을 것 같습니다. 

마찬가지로, `Notification.passToSelectedCategories` 도, `WriteCategoryTableCell`과, `WriteViewModel` 두 곳에서 옵저빙하고 있는데, 위와 같은 이유로 `WriteViewModel`이 전담하도록 하는 편이 코드 복잡성을 더 낮춰주고, `ViewModel`객체의 목적과도 부합한다고 생각합니다. 


4. 그리고 앞으로 추가하실 ( 위에 언급한 ) , 이미지 삭제후 `WrtieViewModel`에도 반영이 되도록 하는 로직과 더불어, `Notificaiton.removeSelectedCategory`에 대한 로직도 `WriteViewModel`이 전담하도록 하면 좋을 것 같습니다. 
지금 `removeSelectedCategory`예로들면,  삭제시, `removeSelectedCategory`를 `post`하고, `category`를 삭제하는 형식인데, 
이것보다, `WriteViewModel`이 데이터를 모두 전담하여, `post`만 호출하고, `WriteViewModel`이 `category`를 삭제하고, 이를 옵저빙하여, `WriteCategoryTableCell`를 다시 그리는 방법이 더 좋을 것 같습니다.

jwonyLee committed:
[#156] ref: 피드백 반영
- PhotosViewController deinit할 때 cache 중지
- Notification.passSelectAssets를 WriteViewModel 전담하도록 변경
- WriteViewModel.executeOriginalImages, requestThumbnailImages 메서드 request**로 통일
- WriteViewModel.cellIdentifier() 의존성 제거
- WriteViewModel.createAttachment 수정
- PageType 주석 추가
- Notification.passToSelectedCategories를 WriteViewModel이 전담하도록 변경
- Notification.Name.removeSelectedCategory 삭제
- WriteImageTAbleCell 에서 사진 삭제 시, WriteViewModel에 삭제 전달

jwonyLee commented:
말씀해주신 피드백 내용 반영했습니다. 
- PhotosViewController deinit할 때 cache 중지
- Notification.passSelectAssets를 WriteViewModel 전담하도록 변경
- WriteViewModel.executeOriginalImages, requestThumbnailImages 메서드 request**로 통일
- WriteViewModel.cellIdentifier() 의존성 제거
- WriteViewModel.createAttachment 수정
- PageType 주석 추가
- Notification.passToSelectedCategories를 WriteViewModel이 전담하도록 변경
- Notification.Name.removeSelectedCategory 삭제
- WriteImageTableCell 에서 사진 삭제 시, WriteViewModel에 삭제 전달

gustn3965 commented on ThingLog/ViewController/CategoryViewController.swift:
```diff
@@ -86,9 +86,10 @@ final class CategoryViewController: UIViewController {
-                                                userInfo: [Notification.Name.passToSelectedCategories: self.convertSelectedIndexPathToCategory()])
+                                                userInfo: [Notification.Name.passToSelectedCategories: category])
                 self.coordinator?.back()
             }.disposed(by: disposeBag)
```
이 코드에 해당하는 부분은 아니지만, 코멘트 드릴 부분이 CategoryViewController안에 있는 코드이기 때문에 여기서 작성하겠습니다.

카테고리뷰컨트롤러에서,  ( 카테고리가 이미 여러개 있는 경우를 가정 )  가장첫번째 카테고리를 선택한다음,  단어순이 가장 빠른 단어로 카테고리를 새로 만들면, 선택된 셀이 방금 만든 셀로 선택됩니다. 
* 지금 현재 카테고리 선택한 경우들을  `selectedCategoryIndexPaths` 의 `indexPath`배열로 저장하고 있고, 이를 통해 체크여부를 판단하는데, `indexPath`보다는, `CategoryEntity`를 이용하여 저장하여, `identifier`를 이용하여 체크여부를 판단하면 될 것 같습니다. 


------------ 
추가적으로 다른 파일에서도 최대한 지원님이 `// MARK`주석을 잘 구분지어 주시고 계시는데, 가끔씩 몇개는 빠져있습니당 🥲 여기서 `viewDidLoad`사이에 `// MARK `주석을 넣어주시면 더 편하게 볼 수 있을 것 같습니다! 감사합니다!  ( 어느정도 메서드 프로퍼티들이 길어지면 넣어주는게 좋을 것 같습니다 👍 )

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -33,45 +33,103 @@ final class WriteViewModel {
+    /// 썸네일 크기
+    private let thumbnailSize: CGSize = CGSize(width: 100, height: 100)
+    /// 이미지가 저장될 크기
+    private let imageSize: CGSize = CGSize(width: 588, height: 588)
```
이미지 사이즈는 고정적으로 가실건가요??

이전 커밋처럼 디바이스 크기별로 가면 안될까요? 
왜냐하면 포토카드에서 사진을 내보낼 때, 디바이스 크기별로 유동적으로 내보내려고 하는데, 그렇게 잘 되기 위해서는 글쓰기 화면에서 이미지를 저장할 때 부터도 디바이스 크기별로 저장하는게 좋을 것 같아서요!

jwonyLee replied:
반영하겠습니다~

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -109,21 +153,69 @@ extension WriteViewModel {
+    }
+
+    /// 이미지가 선택되어 있는 지 여부를 `isSelectImages` 에 갱신한다.
+    private func bindIsImageSelected() {
```
저는 `bindIsImageSelected()` 보다는 `bindThumbnailImagesSubject()`  가 더 좋을 것 같아요 ~ . 왜냐하면, 실제로 구독하는 객체가 `thumbnailImagesSubject` 이기 때문이고, 기존 메소드명만 보면 드는 생각이 "`isImageSelected`는 `Bool`타입인데, 왜 바인딩한다는 거지? 라는 생각이 들었어요. " 지원님은 어떻게 생각하시나요?!👀

jwonyLee replied:
`isSelectImages` 값을 갱신한다는 데 초점을 두고 메서드명을 지었는데, 현수님 말씀하신 부분이 맞는 거 같습니다!

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -33,45 +33,103 @@ final class WriteViewModel {
+    private(set) var thumbnailImagesSubject: BehaviorSubject<[UIImage]> = BehaviorSubject<[UIImage]>(value: [])
+    private(set) var categorySubject: BehaviorSubject<[Category]> = BehaviorSubject<[Category]>(value: [])
     private let disposeBag: DisposeBag = DisposeBag()
 
```
지원님이 전체적으로` // MARK `주석을 잘 달고 계신데 가끔씩 몇개 깜빡하시는 것 같아요 ㅎㅎ 
코드가 전체적으로 길어지다보니 여기에도 `// MARK `주석 달아주시면 보는데 있어서 더 편할 것 같습니당! 🙏🏻

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -33,45 +33,103 @@ final class WriteViewModel {
+
+        switch pageType {
+        case .bought:
+            price = Int(typeValues[1]?.filter("0123456789".contains) ?? "") ?? 0
+            purchasePlace = typeValues[2] ?? ""
+        case .wish:
+            price = Int(typeValues[1]?.filter("0123456789".contains) ?? "") ?? 0
+            purchasePlace = typeValues[2] ?? ""
```
이거 계속 궁금했었는데 왜 0123456789 를 체크하나요??👀

jwonyLee replied:
WriteViewController의 가격 필드에서 숫자를 입력하면 세 자리마다 `,` 와 가장 끝에 ` 원`을 포함해서 사용자한테 보여주고 있는데요, Core Data에는 price를 Int로 저장하고 있기 때문에 가격 필드에서 가져온 값을 그대로 사용할 수 없어서 숫자만 필터링하고 있습니당

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -33,45 +33,103 @@ final class WriteViewModel {
+                    attachments: attachments)
+    }
+
+    /// originalImages로 [Attachment] 를 생성한다.
+    private func createAttachment() -> [Attachment] {
+        var attachments: [Attachment] = []
+        for index in 0..<originalImages.count {
+            if let thumbnail: UIImage = try? thumbnailImagesSubject.value()[index] {
+                let attachment: Attachment = Attachment(thumbnail: thumbnail,
+                                                        imageData: .init(originalImage: originalImages[index]))
+                attachments.append(attachment)
+            }
+        }
+        return attachments
```
WriteViewModel - createAttachment() 에서 index out of range 에러가 나고있습니다.
* 상황 
    * 우선 이미지를 2개 선택하여 돌아온다음,
    * 다시 이미지를 1개만 선택하여 돌아온 다음 저장하는 경우 
*  originalImages 가 현재 3개로 되어있기 때문입니다.


해결책은 아래에 적어놨습니다~!

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -109,21 +153,69 @@ extension WriteViewModel {
+                    self.originalImages.append(image)
                 }
             }
         }
```
위에 createAttachment에서 에러가 발생하는 이유를 여기서 해결할 수 있을 것 같습니다! 
thumbnail 이미지를 만드는 것 처럼 빈 이미지 배열을 이용하여 넣는 방식으로요~
```swift
     /// 파라미터로 받은 `PHAsset`을 `UIImage`로 변환하여 originalImages에 저장한다. 비동기로 동작한다.
    /// - Parameter assets: UIImage로 변환할 [PHAsset]
    private func requestOriginalImages(with assets: [PHAsset]) {
        let options: PHImageRequestOptions = PHImageRequestOptions()
        options.isSynchronous = true
        var images: [UIImage] = []
        DispatchQueue.global().async { [weak self] in
            guard let self = self else { return }
            assets.forEach { asset in
                asset.toImage(targetSize: self.imageSize, options: options) { image in
                    guard let image = image else { return }
                    images.append(image)
                    if images.count == assets.count {
                        self.originalImages = images
                    }
                }
            }
        }
    }
```

gustn3965 commented on ThingLog/ViewModel/WriteViewModel.swift:
```diff
@@ -109,21 +153,69 @@ extension WriteViewModel {
+        }
+
+        thumbnailImagesSubject.onNext(images)
+    }
```
생각해보니, asset.toImage의 클로저는 비동기적으로 발생하므로, 안전하게 아래와 같이 추가하는게 더 좋을 것 같습니다.
지금은 이미지가 작아서 모두 다 추가되겠지만, 안될수도 있기 때문입니다.  

```swift
        assets.forEach { asset in
            asset.toImage(targetSize: self.thumbnailSize, options: options) { image in
                guard let image = image else { return }
                images.append(image)
                if images.count == assets.count {
                    self.thumbnailImagesSubject.onNext(images)
                }
            }
        }
```

gustn3965 reviewed (approved):
`WriteViewModel`만으로 이용하니 코드가 굉장히 많이 깔끔해진 것 같습니다! 
그리고 subject를 이용하여 다른 셀들도 이를 구독하게 하신것도 굉장히 좋은 것 같습니다!!👍

코멘트들 확인하시고, 문제점들 해결하신다음 테스트하시고 `merge`하시면 될 것 같습니다.

jwonyLee committed:
[#156] ref: 피드백 반영
- CategoryViewController 에서 카테고리가 선택되어 있는 경우에 카테고리를 새로 생성하면 선택된 셀이 갱신되지 않은 문제 해결
- 저장할 이미지 크기를 디바이스 Width로 변경
- bindIsImageSelected() -> bindThumbnailImagesSubject()
- 주석 추가
- createAttachments() 수정
- requestThumbnailImages(with:) 수정

jwonyLee committed:
[#156] ref: 선택한 카테고리 저장 방식 변경

기존에 Set<IndexPath>, Set<UUID?>를 별도로 저장하고 있던 방식에서 [CategoryEntity] 만 이용하는 방식으로 변경
